### ISSUE TITLE
perltrap.pod needs updating

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12901]
Migrated from [rt.perl.org#117507](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117507) (status was 'open')

Searchable as RT117507$

**Comment by p5pRT** [comment]
### From @Hugmeir

#### Created by fraserb@gmail.com



perltrap\.pod is supposed to be a list of traps for both perl programmers
and programmers coming to perl from other languages\. However\, it hasn't
really been updated in the last decade\, and is in dire need of an overhaul\.
Currently it only lists a handful of traps for C\, awk\, bash\, sed\, and Perl
itself\. All of these could be cleaned up and expanded\, and also sections
for $LANGUAGE\_OF\_CHOICE could be added\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.16.2:

Configured by hugmeir at Tue Nov 20 17:20:00 ART 2012.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:

  Platform:
    osname=linux, osvers=3.5.0-18-generic,
archname=x86_64-linux-thread-multi
    uname='linux naw 3.5.0-18-generic #29-ubuntu smp fri oct 19
10:26:51 utc 2012 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de
-Dprefix=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2 -DDEBUGGING
-Dusethreads -Doptimize=-g -O0 -ggdb3 -Uversiononly -Accflags=-Wall
-Wextra -Aeval:scriptdir=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra
-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O0 -ggdb3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -O0 -ggdb3
-L/usr/local/lib -fstack-protector'

Locally applied patches:



@INC for perl 5.16.2:
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/x86_64-linux-gnu-thread-multi
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/x86_64-linux-thread-multi
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5

/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2

/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2
    .


Environment for perl 5.16.2:
    HOME=/home/hugmeir
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/home/hugmeir/.rbenv/shims:/home/hugmeir/.rbenv/bin:/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin:/home/hugmeir/.rbenv/shims:/home/hugmeir/.rbenv/bin:/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL5LIB=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/x86_64-linux-gnu-thread-multi:/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5
    PERLBREW_BASHRC_VERSION=0.46
    PERLBREW_HOME=/home/hugmeir/.perlbrew
    PERLBREW_LIB=all
    PERLBREW_MANPATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/man:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/man
    PERLBREW_PATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin
    PERLBREW_PERL=perl-5.16.2
    PERLBREW_ROOT=/home/hugmeir/perl5/perlbrew
    PERLBREW_VERSION=0.46
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MB_OPT=--install_base /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MM_OPT=INSTALL_BASE=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From tchrist@perl.com

Brian Fraser \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote
&nbsp;   on Sat\, 06 Apr 2013 17&#8203;:07&#8203;:36 PDT&#8203;: 

> \[stuff\]

Huh\, could've sworn I'd done that already\, but maybe it only
made it into the Camel Book instead of also making it to perltrap\.pod
as had been the intent\.

\-\-tom

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Leont

On Sun\, Apr 7\, 2013 at 3&#8203;:50 PM\, Tom Christiansen \<tchrist@&#8203;perl\.com> wrote&#8203;:

> Huh\, could've sworn I'd done that already\, but maybe it only
> made it into the Camel Book instead of also making it to perltrap\.pod
> as had been the intent\.
>

You posted something on the list\, see
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2012/01/msg182906\.html\,
but it get enough of a follow\-up it would seem\.

Leon

**Comment by p5pRT** [comment]
### From tchrist@perl.com

Leon wrote&#8203;:

>> Huh\, could've sworn I'd done that already\, but maybe it only
>> made it into the Camel Book instead of also making it to perltrap\.pod
>> as had been the intent\.

> You posted something on the list\, see
> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2012/01/msg182906\.html\,
> but it get enough of a follow\-up it would seem\.

Probably my fault; I was quite distracted at the time\.  That pod was an
early phase of it; it isn't exactly what got published\.

I'd really like for people who know those languages better to hack at it\,
both to remove and to add to it\.  

Too much is seen through tchrist\-colored blinders\, if you know what I mean\. 
In particular\, Java to a much greater extent and Python to a somewhat lesser 
one both left me with a bad taste in my mouth\, and this shines through too 
strongs\, I think\.

I trimmed them down a good bit for publication\, but probably not enough\.

In other words\, they're in some ways written from exactly the wrong
point of view\.  Instead of a Java/Python/Ruby programmer trying 
to come to grips with Perl's traps\, it's actually too much the other
way around\.  Java's annoying object model and both Java and Python's
poor Unicode handling both in and out of regexes kept annoying the dickens
out of me when I would try to use them\.

But you write what you know\, not what you don't\, so that's what you got\.

Feel free to take\, leave\, fold\, spindle\, or mutilate in any way that
you see fit\.  I promise not to complain\.

\-\-tom

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Apr 06 17&#8203;:07&#8203;:36 2013\, Hugmeir wrote&#8203;:
> This is a bug report for perl from fraserb@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> perltrap\.pod is supposed to be a list of traps for both perl
>    programmers
> and programmers coming to perl from other languages\. However\, it
>    hasn't
> really been updated in the last decade\, and is in dire need of an
>    overhaul\.
> Currently it only lists a handful of traps for C\, awk\, bash\, sed\, and
>    Perl
> itself\. All of these could be cleaned up and expanded\, and also
>    sections
> for $LANGUAGE\_OF\_CHOICE could be added\.

I can add a JavaScript section\.  I have one almost written; it just
needs a bit of polishing\.  The existing perltrap documententation uses
alliteration for each language \(cerebral C programmers\, seasoned sed
programmers\, etc\.\)\.  What should I use?  Judicious?  Jabbering? :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jun 03 18&#8203;:48&#8203;:54 2013\, sprout wrote&#8203;:
> On Sat Apr 06 17&#8203;:07&#8203;:36 2013\, Hugmeir wrote&#8203;:
> > This is a bug report for perl from fraserb@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > perltrap\.pod is supposed to be a list of traps for both perl
> >    programmers
> > and programmers coming to perl from other languages\. However\, it
> >    hasn't
> > really been updated in the last decade\, and is in dire need of an
> >    overhaul\.
> > Currently it only lists a handful of traps for C\, awk\, bash\, sed\, and
> >    Perl
> > itself\. All of these could be cleaned up and expanded\, and also
> >    sections
> > for $LANGUAGE\_OF\_CHOICE could be added\.
> 
> I can add a JavaScript section\.

I have added a JavaScript section in commit 9b12f83\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @ap

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-04 03&#8203;:50\]&#8203;:
> The existing perltrap documententation uses alliteration for each
> language \(cerebral C programmers\, seasoned sed programmers\, etc\.\)\.
> What should I use \[for Javascript\]? Judicious? Jabbering? :\-\)

Junior? Jaded? Jealous? Joyless? Jejune? Journeying? Jocund? Joyous?
Jubilant? Jaunty? Jingling? Jazzy? Juicy? Juvenile? Jiggly? Jittery?
Jumpy? Jinxed? Judgemental? Jeering? Jerkish? Jaywalking? Jewish?
Jeopardous? Junctive? Jumbling? Jocular? Jovial? Jolly? Joking

aside\, I guess of all of these that are even contenders\, “jaded” comes
closest to the theme in seasoned/cerebral/practising\, but I don’t know
if that makes it suitable\. I would have liked to come up with something
connoting “fashion\-conscious” but nothing came to mind\, alas\.

\-\- 
\*AUTOLOAD=\*\_;sub \_\{s/&#8203;::\(\[^&#8203;:\]\*\)$/print$1\,\("\,$\\/"\," "\)\[defined wantarray\]/e;chop;$\_\}
&Just\->another\->Perl\->hack;
\#Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jun 03 21&#8203;:38&#8203;:49 2013\, aristotle wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-04
03&#8203;:50\]&#8203;:
> > The existing perltrap documententation uses alliteration for each
> > language \(cerebral C programmers\, seasoned sed programmers\, etc\.\)\.
> > What should I use \[for Javascript\]? Judicious? Jabbering? :\-\)
> 
> Junior? Jaded? Jealous? Joyless? Jejune? Journeying? Jocund? Joyous?
> Jubilant? Jaunty? Jingling? Jazzy? Juicy? Juvenile? Jiggly? Jittery?
> Jumpy? Jinxed? Judgemental? Jeering? Jerkish? Jaywalking? Jewish?
> Jeopardous? Junctive? Jumbling? Jocular? Jovial? Jolly? Joking
> 
> aside\, I guess of all of these that are even contenders\, “jaded” comes
> closest to the theme in seasoned/cerebral/practising\,

But jaded means ‘bored or lacking enthusiasm\, typically after having had
too much of something’\,\* which is too negative\, I think\.

> but I don’t know
> if that makes it suitable\. I would have liked to come up with something
> connoting “fashion\-conscious” but nothing came to mind\, alas\.

I used ‘judicious’\. :\-\)

\* Oxford Dictionary of English 3rd edition 
&nbsp;  Copyright © 2010 by Oxford University Press

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @ap

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-04 07&#8203;:40\]&#8203;:
> But jaded means ‘bored or lacking enthusiasm\, typically after having
> had too much of something’\,\* which is too negative\, I think\.

The way I see the word used it means something like “blasé after too
much experience”\, someone who remains detached because he’s seen it
all and was disillusioned\. Implied by that is a sense of significant
experience\, hence my comment\.

The man’s or the lady’s disillusionment\, obviously\, would stem from the
language\. :\-\)  \(And let it be said here&#8203;: I like the language\. I enjoy
it in spite of some bizarre design choices in very everyday parts of the
language\. Maybe it is I who is a jaded then… sometimes\, at least\.\)

But yes\, too negative for a perldoc\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jun 04 05&#8203;:57&#8203;:53 2013\, aristotle wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-04
07&#8203;:40\]&#8203;:
> > But jaded means ‘bored or lacking enthusiasm\, typically after having
> > had too much of something’\,\* which is too negative\, I think\.
> 
> The way I see the word used it means something like “blasé after too
> much experience”\, someone who remains detached because he’s seen it
> all and was disillusioned\. Implied by that is a sense of significant
> experience\, hence my comment\.
> 
> The man’s or the lady’s disillusionment\, obviously\, would stem from the
> language\. :\-\)  \(And let it be said here&#8203;: I like the language\. I enjoy
> it in spite of some bizarre design choices in very everyday parts of the
> language\.

I know this is going off\-topic a bit\, but I feel the same way\.  The most
bizarre case I’ve found is the behaviour of with\(\.\.\.\) var foo = \.\.\.\.  
The variable declaration creates a global \(or function\) variable at
compile time \(or function entry\)\, but at run time it may refer to an
object in the with\(\) scope\!

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Apr 06 17&#8203;:07&#8203;:36 2013\, Hugmeir wrote&#8203;:
> This is a bug report for perl from fraserb@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> perltrap\.pod is supposed to be a list of traps for both perl
>    programmers
> and programmers coming to perl from other languages\. However\, it
>    hasn't
> really been updated in the last decade\, and is in dire need of an
>    overhaul\.
> Currently it only lists a handful of traps for C\, awk\, bash\, sed\, and
>    Perl
> itself\. All of these could be cleaned up and expanded\, and also
>    sections
> for $LANGUAGE\_OF\_CHOICE could be added\.
> 

hugmeir\,

I really appreciate the sentiment behind this ticket\.  However\, it's
formulated without specific criteria for closing the ticket\.  In other
words\, how will we know when we have sufficiently improved
pod/perltrap\.pod such that this ticket could be closed\.

Would it be better if we created separate RT tickets for each language
that needed some cleanup\, with pointers to specific things that need
cleaning up?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @xdg

On Tue\, Jun 4\, 2013 at 12&#8203;:47 PM\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> I really appreciate the sentiment behind this ticket\.  However\, it's
> formulated without specific criteria for closing the ticket\.

Excellent point\, Jim\.

> Would it be better if we created separate RT tickets for each language
> that needed some cleanup\, with pointers to specific things that need
> cleaning up?

\+1

\-\-
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jun 04 09&#8203;:47&#8203;:59 2013\, jkeenan wrote&#8203;:
> On Sat Apr 06 17&#8203;:07&#8203;:36 2013\, Hugmeir wrote&#8203;:
> > This is a bug report for perl from fraserb@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > perltrap\.pod is supposed to be a list of traps for both perl
> >    programmers
> > and programmers coming to perl from other languages\. However\, it
> >    hasn't
> > really been updated in the last decade\, and is in dire need of an
> >    overhaul\.
> > Currently it only lists a handful of traps for C\, awk\, bash\, sed\, and
> >    Perl
> > itself\. All of these could be cleaned up and expanded\, and also
> >    sections
> > for $LANGUAGE\_OF\_CHOICE could be added\.
> > 
> 
> hugmeir\,
> 
> I really appreciate the sentiment behind this ticket\.  However\, it's
> formulated without specific criteria for closing the ticket\.  In other
> words\, how will we know when we have sufficiently improved
> pod/perltrap\.pod such that this ticket could be closed\.
> 
> Would it be better if we created separate RT tickets for each language
> that needed some cleanup\, with pointers to specific things that need
> cleaning up?

Since the Camel has these\, we probably ought to include Java\, Python and
Ruby\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @Leont

On Tue\, Jun 4\, 2013 at 10&#8203;:08 PM\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> Would it be better if we created separate RT tickets for each language
>> that needed some cleanup\, with pointers to specific things that need
>> cleaning up?
>
> Since the Camel has these\, we probably ought to include Java\, Python and
> Ruby\.

Also\, we should really include php in that list\.

Leon
